### ISSUE TITLE
Implementation Doubts

### ISSUE THREAD

**Issue by shubhamag97** [openai/DALL-E#28]
Although this codebase is for the vae part. would appreciate if you could help in understanding few components of the transformer part also.
In the paper and blog released, you mentioned that you use Child et al paper. Can you elaborate on what you use as the block size? 8/16/32
If we use a block size of 16 for example then how do you implement the convolutional kernel, it has gaps of only 1 block but if you have sparse block of size 16 then it doesn't make sense.

Also, when you are training the gpt style model. Even though the loss and perplexity reduce, how do you identify when the perplexity/loss value of the 1.2B parameter model is sufficient? like is a loss of 4/5 good or should it be <1.

**Comment by adityaramesh** [comment]
I used a block size of 32. Blocksparse transformer has a two-level layout: the top-level layout describes which 32x32 tiles in the attention mask are enabled, and the bottom-level layout described the sparsity pattern within the 32x32 block of the attention mask with index (i, j). So any conceivable attention pattern can be implemented, and this is independent of the block size -- it might be worth making the block size smaller in cases where more blocks in the top-level layout can be disabled, since this saves compute.

Regarding scaling, I worked out the scaling laws by training a series of smaller models and made sure that the gains from doubling the model size were consistent.

**Comment by shubhamag97** [comment]
Just a follow-up question. 
As per fig 10 of the paper, there is only pos embedding in the text part and row/col embedding for the image part. Why is there a need for two different positional embedding structures? Can't we just flatten out the image (basically the 1024 tokens) and expect the attention to take care of it. Does using row/col embedding after 256 given a boost in performance over positional embedding till 1024.

Does using 16384 tokens instead of 50257 help for text embedding help with the training somehow?

Also, how do you initialize the text/image tokens? I'm assuming it's done randomly and then learned by the model.
 Does using any pre-trained models like gpt-2 (with hidden dimension 3968) for text and VQVAE tokens for image tokens as initialization. Can the use of pre-trained models for initialization help?

**Comment by adityaramesh** [comment]
> Can't we just flatten out the image (basically the 1024 tokens) and expect the attention to take care of it. Does using row/col embedding after 256 given a boost in performance over positional embedding till 1024.

Sparse transformer reports a small boost from using factored embeddings and my model is based on the one reported there, so I just borrowed this implementation detail. I didn't do a careful ablation for this.

> Does using 16384 tokens instead of 50257 help for text embedding help with the training somehow?

The limited vocabulary size was purely for memory reasons due to having to train a 12b model on 16 GB V100 GPUs.

> Also, how do you initialize the text/image tokens? I'm assuming it's done randomly and then learned by the model.

Yes, this was done randomly.

> Does using any pre-trained models like gpt-2 (with hidden dimension 3968) for text and VQVAE tokens for image tokens as initialization. Can the use of pre-trained models for initialization help?

I never tried this but I think it'd be interesting to see!

**Comment by shubhamag97** [comment]
Just a follow-up question,
You mentioned the top and bottom layer in sparsity of 32x32. In case of dalle the input dimension is 1280. with block size of 32, it is 40x40. Do you divide the attention matrix into text attention and image attention and then add them after applying separately? If so can you elaborate on any optimization or methods you use there?
 or do you use one attention matrix using block-sparse? If so then how do you handle two different bottom level attention in qkv multiplication
